### PR TITLE
Bot basis creation

### PR DESCRIPTION
**Pull request by vladyslav-pustovalov**

### CODE REVIEW

File: src/main/java/project/bot/config/BotConfig.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package project.bot.config;
+
+/**
+ * @author Vladyslav Pustovalov
+ * class which contains Configuration bot data
+ */
```
Comment (Dmytro-Vusyk) [inline src/main/java/project/bot/config/BotConfig.java]: comment should be before class declaration

File: src/main/java/project/bot/config/BotConfig.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+@Configuration
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class BotConfig {
+
+    final  String botName = System.getenv("botName");
+    final  String botToken = System.getenv("botToken");
+
+    final DefaultBotOptions botOptions = new DefaultBotOptions();
+
```
Comment (Dmytro-Vusyk) [inline src/main/java/project/bot/config/BotConfig.java]: remove blank line

File: src/main/java/project/bot/service/KeyboardSetter.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class KeyboardSetter {
+
+    /**
+     * Method which initializes default keyboard for the bot
+     */
+    public ReplyKeyboardMarkup setDefaultKeyboard() {
```
Comment (Dmytro-Vusyk) [inline src/main/java/project/bot/service/KeyboardSetter.java:19]: variables must be declared first. Than should be actions

File: src/main/java/project/bot/service/MessageSender.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
+package project.bot.service;
+
+import lombok.AccessLevel;
+import lombok.experimental.FieldDefaults;
+import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
+
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class MessageSender {
```
Comment (Dmytro-Vusyk) [inline src/main/java/project/bot/service/MessageSender.java:9]: all hardcoded strings should be declared as constants

File: src/main/java/project/bot/service/TenderParserServiceBot.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+import org.telegram.telegrambots.meta.api.objects.Update;
+import org.telegram.telegrambots.meta.exceptions.TelegramApiException;
+import project.bot.config.BotConfig;
+
+@Component
+@Slf4j
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class TenderParserServiceBot extends TelegramLongPollingBot{
+    final BotConfig config = new BotConfig();
+
```
Comment (Dmytro-Vusyk) [inline src/main/java/project/bot/service/TenderParserServiceBot.java]: remove blank lines

File: src/main/java/project/bot/service/KeyboardSetter.java (hunk 3 above)
Comment (vladyslav-pustovalov) [reply src/main/java/project/bot/service/KeyboardSetter.java:19]: but there are no variables in this class, all of them are declared in the method

Comment (Dmytro-Vusyk) [reply src/main/java/project/bot/service/KeyboardSetter.java:19]: The comment is left for the method, not class

File: src/main/java/project/bot/service/MessageSender.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    static final String AVAILABLE_SITES_TEXT = "Available sites for subscription";
+    static final String HELP_TEXT = "Help text";
+    static final String DEFAULT_TEXT = "Sorry, this command is not supported!\nPlease use the Keyboard buttons!";
+    static final String WARN_TEXT = "You are trying to make bad things!\nPlease use the Keyboard buttons!";
+    static final KeyboardSetter keyboard = new KeyboardSetter();
+
+    /**
+     * Method which sends the welcome text to a user
+     */
+    public SendMessage startMessage(long chatId) {
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/MessageSender.java]: Maybe another method's name suit better - e.g. sendWelcomeMessage() ?

File: src/main/java/project/bot/service/MessageSender.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        message.setChatId(chatId);
+        message.setText(WELCOME_TEXT);
+        message.setReplyMarkup(keyboard.setDefaultKeyboard());
+        return message;
+    }
+
+    /**
+     * Method which sends to a user the list with they subscription
+     */
+    public SendMessage userSubscriptionsList(long chatId) {
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/MessageSender.java]: sendUserSubscriptionsList() ?

and below:
sendAvailableSitesList()
send...
send...

File: src/main/java/project/bot/service/MessageSender.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        message.setChatId(chatId);
+        message.setText(HELP_TEXT);
+        message.setReplyMarkup(keyboard.setDefaultKeyboard());
+        return message;
+    }
+
+    /**
+     * Method which sends default answer on not supported commands
+     */
+    public SendMessage defaultMessage(long chatId) {
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/MessageSender.java]: sendUnsupportedCommandWarning() ?

File: src/main/java/project/bot/service/MessageSender.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        message.setChatId(chatId);
+        message.setText(DEFAULT_TEXT);
+        message.setReplyMarkup(keyboard.setDefaultKeyboard());
+        return message;
+    }
+
+    /**
+     * Method which sends default answer on not supported update date format
+     */
+    public SendMessage warnMessage(long chatId) {
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/MessageSender.java]: sendUnsupportedUpdateDataFormatWarning() ?

File: src/main/java/project/bot/config/BotInitializer.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+ */
+@Component
+@Slf4j
+@FieldDefaults(level = AccessLevel.PRIVATE)
+public class BotInitializer {
+
+    final BotConfig config = new BotConfig();
+
+    final TenderParserServiceBot bot = new TenderParserServiceBot(config.getBotOptions(), config.getBotToken());
+
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/config/BotInitializer.java:21]: Why not private ?

File: src/main/java/project/bot/service/MessageSender.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+public class MessageSender {
+
+    static final String WELCOME_TEXT = "Welcome text";
+    static final String USER_SUBSCRIPTION_TEXT = "User's subscription list";
+    static final String AVAILABLE_SITES_TEXT = "Available sites for subscription";
+    static final String HELP_TEXT = "Help text";
+    static final String DEFAULT_TEXT = "Sorry, this command is not supported!\nPlease use the Keyboard buttons!";
+    static final String WARN_TEXT = "You are trying to make bad things!\nPlease use the Keyboard buttons!";
+    static final KeyboardSetter keyboard = new KeyboardSetter();
+
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/MessageSender.java:17]: You don't use them in static methods.
Why "static" ?

File: src/main/java/project/bot/service/TenderParserServiceBot.java [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                case "Available tender sites for subscription" -> {
+                    try {
+                        execute(messageSender.availableSitesList(chatId));
+                        log.info("Available sites list was sent to user " + userName);
+                    } catch (TelegramApiException e) {
+                        log.error(errorOccurred + e);
+                    }
+                }
+
+                case "Help instructions" -> {
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/TenderParserServiceBot.java]: "My subscription list"
"Available tender sites for subscription"
"Help instructions"

They are too long(

Should be only 1 short word w/o CAPS:
subs
sites
help

File: src/main/java/project/bot/service/TenderParserServiceBot.java [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+     * Method which gets updates from users and sends to them an answer
+     */
+    @Override
+    public void onUpdateReceived(Update update) {
+        String userName = update.getMessage().getChat().getUserName();
+        String messageText = update.getMessage().getText();
+        long chatId = update.getMessage().getChatId();
+        String errorOccurred = "Error occurred: ";
+
+        if (!update.hasMessage() || !update.getMessage().hasText()) {
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/TenderParserServiceBot.java:43]: if (update.hasMessage() && update.getMessage().hasText()) {
    switch(messageText) {...}
} else {
    log.warn("...
}

File: src/main/java/project/bot/config/BotInitializer.java (hunk 10 above)
Comment (vladyslav-pustovalov) [reply src/main/java/project/bot/config/BotInitializer.java:21]: they are private, with Lombok annotation @FieldDefaults

File: src/main/java/project/bot/service/TenderParserServiceBot.java (hunk 12 above)
Comment (vladyslav-pustovalov) [reply src/main/java/project/bot/service/TenderParserServiceBot.java]: I thought it looks clear and beautiful)

File: src/main/java/project/bot/service/TenderParserServiceBot.java (hunk 13 above)
Comment (vladyslav-pustovalov) [reply src/main/java/project/bot/service/TenderParserServiceBot.java:43]: I event don't know, @Dmytro-Vusyk  what will you say, which style will be better?

File: src/main/java/project/bot/service/TenderParserServiceBot.java (hunk 12 above)
Comment (MykolaLisnyi) [reply src/main/java/project/bot/service/TenderParserServiceBot.java]: you can create constants with those names

File: src/main/java/project/bot/service/TenderParserServiceBot.java [hunk 14]
```diff
@@ -38,59 +36,43 @@ public void onUpdateReceived(Update update) {
...
-                        execute(messageSender.sendUserSubscriptionsList(chatId));
-                        log.info("Subscription list was sent to user " + userName);
-                    } catch (TelegramApiException e) {
-                        log.error(errorOccurred + e);
-                    }
+                    sendMessageToUser(messageSender.sendUserSubscriptionsList(chatId));
+                    log.info("Subscription list was sent to user " + userName);
+
                 }
                 case "Available tender sites for subscription" -> {
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/TenderParserServiceBot.java]: I don't want to type so long messages(((
I prefere short ones.

But constats' names can be long)

File: src/main/java/project/bot/service/MessageSender.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    /**
+     * Method which sends the welcome text to a user
+     */
+    public SendMessage sendWelcomeMessage(long chatId) {
+        SendMessage message = new SendMessage();
+        message.setChatId(chatId);
+        message.setText(welcomeText);
+        message.setReplyMarkup(keyboard.setDefaultKeyboard());
+        return message;
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/MessageSender.java:26]: What about using method chaining for all methods below?

```
public SendMessage sendWelcomeMessage(long chatId) {
return new SendMessage().setCharId(chatId)
                        .setText(welcomeText)
                        .setReplyMarkup(keyboard.setDefaultKeyboard());
}
```

File: src/main/java/project/bot/service/TenderParserServiceBot.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+    /**
+     * Method which gets updates from users and sends to them an answer
+     */
+    @Override
+    public void onUpdateReceived(Update update) {
+        String userName = update.getMessage().getChat().getUserName();
+        String messageText = update.getMessage().getText();
+        long chatId = update.getMessage().getChatId();
+
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/TenderParserServiceBot.java:42]: Maybe ?

```
Message message = update.getMessage();
String userName = message.getChat().getUserName();
String messageText = message.getText();
long chatId = message.getChatId();
```

File: src/main/java/project/bot/service/TenderParserServiceBot.java [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+                default -> {
+                    sendMessageToUser(messageSender.sendUnsupportedCommandWarning(chatId));
+                    log.info("Unsupported command warning was sent to user " + userName);
+                }
+            }
+        }
+    }
+
+    private void sendMessageToUser(SendMessage message) {
+        try {
```
Comment (MykolaLisnyi) [inline src/main/java/project/bot/service/TenderParserServiceBot.java:76]: Variable "sendMessage" looks better.

And you clearly understand what should do method "execute"
`execute(sendMessage); `

Comment (vladyslav-pustovalov) [reply src/main/java/project/bot/service/TenderParserServiceBot.java:76]: I guess yes, it sends any message from bot to user vie chat id

File: src/main/java/project/bot/service/TenderParserServiceBot.java (hunk 16 above)
Comment (vladyslav-pustovalov) [reply src/main/java/project/bot/service/TenderParserServiceBot.java:42]: is it really necessary?)

File: src/main/java/project/bot/service/TenderParserServiceBot.java (hunk 14 above)
Comment (vladyslav-pustovalov) [reply src/main/java/project/bot/service/TenderParserServiceBot.java]: you do not need to type it, this case works when you push such button on the keyboard

File: src/main/java/project/bot/service/TenderParserServiceBot.java (hunk 16 above)
Comment (MykolaLisnyi) [reply src/main/java/project/bot/service/TenderParserServiceBot.java:42]: 3 times update.getMessage() 
I think it is DRY)

File: src/main/java/project/bot/service/TenderParserServiceBot.java (hunk 14 above)
Comment (MykolaLisnyi) [reply src/main/java/project/bot/service/TenderParserServiceBot.java]: So why not enums ?

File: src/main/java/project/bot/service/MessageSender.java [hunk 18]
```diff
@@ -0,0 +1,83 @@
+package project.bot.service;
+
+import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
+
+/**
+ * @author Vladyslav Pustovalov
+ * class which consist of answers to a user from the Telegram Bot
+ */
+public class MessageSender {
```
Comment (Dmytro-Vusyk) [inline src/main/java/project/bot/service/MessageSender.java:9]: you can shorten methods to rebese logic from them to private method foo(long id, string text)

File: src/main/java/project/bot/service/TenderParserServiceBot.java [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+        String messageText = message.getText();
+        long chatId = message.getChatId();
+
+        if (!update.hasMessage() || !update.getMessage().hasText()) {
+            log.warn("Unexpected update from user " + userName);
+            sendMessageToUser(messageSender.sendUnsupportedDataWarning(chatId));
+            log.info("Warn message was sent to user " + userName);
+        } else {
+            switch (messageText) {
+                case "/start" -> {
```
Comment (Dmytro-Vusyk) [inline src/main/java/project/bot/service/TenderParserServiceBot.java:49]: please add brake to to the switch case statement

**Review by Dmytro-Vusyk — APPROVED** [review]

**Review by MykolaLisnyi — APPROVED** [review]
ok

**Pull request merged**